google-apis-common/src/auth.rs

```diff
@@ -0,0 +1,163 @@
+//! Authentication for Google API endpoints
+//!
+//! Allows users to provide custom authentication implementations to suite their needs.
+//!
+//! By default, [`GetToken`] is implemented for:
+//! - [`Authenticator`] : An authenticator which supports a variety of authentication methods
+//! - [`String`] : Plain oauth2 token in String format
+//! - [`NoToken`] : No token, used for APIs which do not require a token
+//!
+//! # Usage
+//! [`GetToken`] instances are designed to be used with the Hub constructor provided by the
+//! generated APIs.
+//!
+//! If you intend to use the API libraries on client devices,
+//! [`Authenticator`] supports a variety of client-side authentication methods,
+//! and should be used to provide authentication.
+//!
+//! If you intend to use the API libraries server-side, with server-side client authentication,
+//! use the [`oauth2`] crate and convert the resulting [`AccessToken`] to [`String`].
```

Comment:
I think you're referring specifically to service accounts - essentially, an account with permission to access a specific set of services. I believe service accounts can be given permission to access any Google API, but are intended mainly for developer / backend use.
 
However, I'm specifically referring to accessing APIs like Google Drive on the behalf of an unknown end user from a remote server (server gets token, accesses Google Drive). This use-case in particular is not supported by `yup-oauth2`, and even if it was, it would not be reasonable to support using the `Authenticator` struct or `GetToken` trait.

This is mainly because the server needs to store state about the user session, which requires redirecting the user to predefined callback URLs. `yup-oauth2` handles this by spinning up a server and blocking until the user response is received - entirely inappropriate for production use.

This makes it difficult to use the same credentials for multiple APIs (don't re-auth the user to use Google Photos / Google Drive - doing this manually and passing the shared token to both hubs would work).

Finally, in a server context, you need to define the oauth2 callback URL and the necessary routes to your webserver, and then manually implement the redirects and oauth2 session state (or use a middleware library). This is especially not supported by `yup-oauth2`.